pkg/parser/io_parser.go

```diff
@@ -0,0 +1,152 @@
+
+	var (
+		bufcap    = cap(b)
+		outbuflen = len(p.outputBuffer)
+	)
+
+	// If somehow we have exactly cap(b) in output buffer, just copy and return.
+	if outbuflen == bufcap {
+		copy(b, p.outputBuffer)
+		p.outputBuffer = p.outputBuffer[:0]
+		return bufcap, nil
+	}
+
+	// If we have more than cap(b), copy, resize and return.
+	if outbuflen > bufcap {
+		copy(b, p.outputBuffer[:bufcap])
+
+		// Proper way to resize the slice without losing it's length
+		copy(p.outputBuffer, p.outputBuffer[bufcap:])
+		p.outputBuffer = p.outputBuffer[:outbuflen-bufcap]
+
+		return bufcap, nil
+	}
+
+	// Not enough data left in output buffer, so drain it unconditionally.
+	copy(b[:outbuflen], p.outputBuffer)
+	p.outputBuffer = p.outputBuffer[:0]
+
+	// Remember how much we can write more
+	written := outbuflen
+
+	// Loop until we have either exhausted the parser OR the buffer.
+	for {
+		res, more := <-p.parse.Results()
+		if !more {
+			return written, io.EOF
+		}
+
+		data, err := json.Marshal(res)
+		if err != nil {
+			return written, err
+		}
+
+		var (
+			datalen = len(data)
+			capleft = bufcap - written
+		)
+
+		// We have exactly enough data to fill in the rest of the buffer, so copy and return.
+		if datalen == capleft {
+			copy(b[written:], data)
+			return bufcap, nil
+		}
+
+		// We have more data to write than we can fit in the output.
+		// Copy however many bytes we can from serialized result and store the rest in output buffer.
+		if datalen > capleft {
+			copy(b[written:], data[:capleft])
+			p.outputBuffer = append(p.outputBuffer, data[capleft:]...)
+			return bufcap, nil
+		}
+
+		// We can fit more than we have available for writing.
+		// Write as much as we can and continue to the next iteration.
+		copy(b[written:], data)
+		written = written + datalen
+	}
+}
+
+func (p *IOParser) Write(b []byte) (int, error) {
+	p.inputLock.Lock()
+	defer p.inputLock.Unlock()
+
+	written := 0
+
+	for {
+		buflen := len(b)
+
+		// Quick check for empty buffer.
+		if buflen == 0 {
+			return written, nil
+		}
+
+		// Check if incoming buffer contains a newline,
+		// since it indicates that we can now send the utterance to AST parser.
+		linesep := bytes.IndexRune(b, newline)
+
+		// If incoming bytestring doesn't contain a newline,
+		// then just append it to our "current line" buffer and return.
+		if linesep == -1 {
+			p.inputBuffer = append(p.inputBuffer, b...)
+			return written + buflen, nil
+		}
+
+		// Split the buffer by the newline character.
+		// Head contains the bytes [0 : newlineIndex]
+		// Re-slice the buffer to make sure we are progressively moving forward on it.
+		head := b[:linesep]
+		b = b[linesep+newlineLen:]
+
+		var line string
+		if len(p.inputBuffer) == 0 {
+			line = string(head)
+		} else {
+			// Update input buffer, get the line to be parsed and zero out the input buffer.
+			p.inputBuffer = append(p.inputBuffer, head...)
+			line = string(p.inputBuffer)
+			p.inputBuffer = p.inputBuffer[:0]
```

Comment:
Shouldn't the input buffer read the rest of `b`? By  [io.Writer](https://golang.org/pkg/io/#Writer) all data should be consumed or error raised.

Reply:
> Shouldn't the input buffer read the rest of b?

No, because the rest of b is read directly to `parser.Parse`. This branch of code is for the case when `Write` is invoked N (N > 1) times, but we only get to the newline char at the last invocation.

Consider the following example:

```
1. parser.Write("str")
2. parser.Write("ing\n")
```

In [1], the bytes are stored in `inputBuffer`, in [2], we detect a newline and then we have to check the `inputBuffer`. In this example, it does contain something, so we concat it with the `head` (i.e. the chars until the newline) and then send the whole line to the parser.

If `b` would've been e.g. `"ing\nstr"` in [2], then the after re-slicing on L132, `b` would only contain `str`, which will be handled on the next loop iteration on L124.

Reply:
Yeah, the same as above, missed the loop 👍